File: PasswordChanger1C/CommonModule.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace PasswordChanger1C
{
    public static class CommonModule
    {
        public static string Format_AdmRole(in bool AdmRole)
        {
            return AdmRole ? "\u2714" : ""; // ✔
        }

        public static string DecodePasswordStructure(in byte[] bytes_Input, ref int KeySize, ref byte[] KeyData)
        {
            short Base = bytes_Input[0];
            KeySize = Base;
            KeyData = new byte[Base];

            bytes_Input.AsMemory(1, Base).CopyTo(KeyData.AsMemory(0, Base));

            int i = Base + 1;
            int j = 1;
            int MaxI = bytes_Input.Length;

            var BytesResult = new byte[MaxI - Base];
            while (i < MaxI)
            {
                if (j > Base)
                {
                    j = 1;
                }

                BytesResult[i - Base - 1] = Convert.ToByte(bytes_Input[i] ^ bytes_Input[j]); // 239 for first

                i++;
                j++;
            }

            return Encoding.UTF8.GetString(BytesResult);
        }

        public static byte[] EncodePasswordStructure(in string Str, int KeySize, in byte[] KeyData)
        {
            var bytes_Input = Encoding.UTF8.GetBytes(Str);
            int Base = KeySize;
            var BytesResult = new byte[bytes_Input.Length + Base];
            BytesResult[0] = Convert.ToByte(Base);

            KeyData.AsMemory(0, Base).CopyTo(BytesResult.AsMemory(1, Base));

            int MaxI = bytes_Input.Length - 1;
            int i = 1;
            int j = 1;

            while (i <= MaxI)
            {
                if (j > Base)
                {
                    j = 1;
                }

                BytesResult[i + Base] = Convert.ToByte(bytes_Input[i - 1] ^ BytesResult[j]);

                i++;
                j++;
            }

            return BytesResult;
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "SecurityIntelliSenseCS:MS Security rules violation", Justification = "SHA-1 hash type")]
        public static string EncryptStringSHA1(in string Str)
        {
            var sha = new SHA1CryptoServiceProvider(); // declare sha as a new SHA1CryptoServiceProvider
            byte[] bytesToHash; // and here is a byte variable
            bytesToHash = Encoding.UTF8.GetBytes(Str); // covert the password into ASCII code
            bytesToHash = sha.ComputeHash(bytesToHash); // this is where the magic starts and the encryption begins
            return Convert.ToBase64String(bytesToHash);
        }

        public static bool IsPasswordHash(in string hashstr)
        {
            // SHA-1 hash size is 160 bit
            const int hash_size = 20; // 160 / 8
            const int base64_size = 28; // ((4 * hash_size / 3) + 3) & ~3;

            if (string.IsNullOrEmpty(hashstr)) return false;

            string base64String = hashstr.Trim('"');
            if (base64String.Length != base64_size
                || base64String.Contains(" ") || base64String.Contains("\t")
                || base64String.Contains("\r") || base64String.Contains("\n"))
                return false;
            try
            {
                byte[] decoded = Convert.FromBase64String(base64String);
                return hash_size == decoded.Length;
            }
            catch (Exception) { }
            return false;
        }

        public static Tuple<string, string> GetPasswordHashTuple(in ParserServices.ParserList AuthStructure)
        {
            string[] result = { "", "" };

            var Hashes = AuthStructure
                 .Where(Item => !Item.IsList)
                 .Select(Item => Item.ToString().Trim('"'))
                 .Where(Item => IsPasswordHash(Item))
                 .Take(2);

            int i = 0;
            foreach (var hash in Hashes)
            {
                result[i] = hash;
                i++;
            }
            return Tuple.Create(result[0], result[1]);
        }

        public static string ReplaceHashes(in string Data, in Tuple<string, string> OldHashes, in Tuple<string, string> NewHashes)
        {
            // Hashes can be with or without double quotes
            static string EnsureDQuotes(string str) => str.StartsWith("\"") && str.EndsWith("\"") ? str : $"\"{str}\"";

            string[] old_arr = { OldHashes.Item1, OldHashes.Item2 };
            string[] new_arr = { NewHashes.Item1, NewHashes.Item2 };
            var old_hashes = old_arr.Select(EnsureDQuotes);
            var new_hashes = new_arr.Select(EnsureDQuotes);
            string OldStr = string.Join(",", old_hashes);
            string NewStr = string.Join(",", new_hashes);
            return Data.Replace(OldStr, NewStr);
        }

        public static Tuple<string, string> GeneratePasswordHashes(in string password)
        {
            return Tuple.Create(EncryptStringSHA1(password), EncryptStringSHA1(password.ToUpper()));
        }

        private static int GetFieldSize(in string FieldType, int FieldLength, int CouldBeNull)
        {
            return FieldType switch
            {
                "B" => FieldLength,
                "L" => 1,
                "N" => (FieldLength + 2) / 2,
                "NC" => FieldLength * 2,
                "NVC" => FieldLength * 2 + 2,
                "RV" => 16,
                "I" => 8,
                "T" => 8,
                "DT" => 7,
                "NT" => 8,
                _ => 0,
            } + CouldBeNull;
        }

        public static int GetFieldSize(in AccessFunctions.TableFields Field)
        {
            return GetFieldSize(Field.Type, Field.Length, Field.CouldBeNull);
        }

        public static void ParseTableDefinition(ref AccessFunctions.PageParams PageHeader)
        {
            var ParsedString = ParserServices.ParsesClass.ParseString(PageHeader.TableDefinition);
            PageHeader.Fields = new List<AccessFunctions.TableFields>();
            int RowSize = 1;
            string TableName = ParsedString[0][0].ToString().Replace("\"", "").ToUpper();
            PageHeader.TableName = TableName;
            foreach (var a in ParsedString[0][2])
            {
                if (!a.IsList)
                {
                    continue;
                }

                var Field = new AccessFunctions.TableFields();
                Field.Name = a[0].ToString().Replace("\"", "");
                Field.Type = a[1].ToString().Replace("\"", "");
                Field.CouldBeNull = Convert.ToInt32(a[2].ToString());
                Field.Length = Convert.ToInt32(a[3].ToString());
                Field.Precision = Convert.ToInt32(a[4].ToString());
                int FieldSize = GetFieldSize(Field);              

                Field.Size = FieldSize;
                Field.Offset = RowSize;
                RowSize += FieldSize;
                PageHeader.Fields.Add(Field);
            }

            PageHeader.RowSize = RowSize;

            // {"Files",118,119,96}
            // Данные, BLOB, индексы

            int BlockData = Convert.ToInt32(ParsedString[0][5][1].ToString());
            int BlockBlob = Convert.ToInt32(ParsedString[0][5][2].ToString());
            PageHeader.BlockData = BlockData;
            PageHeader.BlockBlob = BlockBlob;
        }
    }
}

File: PasswordChanger1C/CommonModule.vb


Imports System.Security.Cryptography
Imports System.Text

Module CommonModule


    Function DecodePasswordStructure(bytes_Input As Byte(), ByRef KeySize As Integer, ByRef KeyData As Byte()) As String

        Dim Base = Convert.ToInt16(bytes_Input.GetValue(0), 10)

        KeySize = Base
        KeyData = New Byte(Base - 1) {}
        For a = 1 To Base
            KeyData(a - 1) = bytes_Input(a)
        Next

        Dim i = Base + 1
        Dim j = 1
        Dim MaxI = bytes_Input.Length

        Dim BytesResult() As Byte = New Byte(MaxI - Base - 1) {}

        While i < MaxI
            If j > Base Then
                j = 1
            End If

            Dim AA = Convert.ToInt16(bytes_Input.GetValue(i), 10)
            Dim BB = Convert.ToInt16(bytes_Input.GetValue(j), 10)
            Dim CC = Convert.ToByte(AA Xor BB) ' 239 for first

            BytesResult.SetValue(CC, i - Base - 1)

            i = i + 1
            j = j + 1

        End While

        Return Encoding.UTF8.GetString(BytesResult)

    End Function


    Function EncodePasswordStructure(Str As String, ByVal KeySize As Integer, ByVal KeyData As Byte()) As Byte()

        Dim bytes_Input = Encoding.UTF8.GetBytes(Str)

        Dim Base = KeySize

        Dim BytesResult() As Byte = New Byte(bytes_Input.Length + Base - 1) {}
        BytesResult.SetValue(Convert.ToByte(Base), 0)
        For ii = 1 To Base
            BytesResult.SetValue(KeyData.GetValue(ii - 1), ii)
        Next

        Dim MaxI = bytes_Input.Length - 1
        Dim i = 1
        Dim j = 1

        While i <= MaxI
            If j > Base Then
                j = 1
            End If

            Dim AA = Convert.ToInt16(bytes_Input.GetValue(i - 1), 10)
            Dim BB = Convert.ToInt16(BytesResult.GetValue(j), 10)
            Dim CC = Convert.ToByte(AA Xor BB)

            BytesResult.SetValue(CC, i + Base)

            i = i + 1
            j = j + 1

        End While

        Return BytesResult

    End Function

    Public Function EncryptStringSHA1(ByVal Str As String) As String

        Dim sha As New SHA1CryptoServiceProvider ' declare sha as a new SHA1CryptoServiceProvider
        Dim bytesToHash() As Byte ' and here is a byte variable

        bytesToHash = System.Text.Encoding.UTF8.GetBytes(Str) ' covert the password into ASCII code

        bytesToHash = sha.ComputeHash(bytesToHash) ' this is where the magic starts and the encryption begins

        Return Convert.ToBase64String(bytesToHash)

        Dim result As String = ""

        For Each b As Byte In bytesToHash
            result += b.ToString("x2")
        Next

        Return result

    End Function

    Sub ParseTableDefinition(ByRef PageHeader As PageParams)

        Dim ParsedString = ParserServices.ParsesClass.ParseString(PageHeader.TableDefinition)

        PageHeader.Fields = New List(Of TableFields)

        Dim RowSize = 1

        Dim TableName = ParsedString.Item(0).Item(0).ToString.Replace("""", "").ToUpper

        PageHeader.TableName = TableName

        For Each a In ParsedString.Item(0).Item(2)
            If Not a.IsList Then
                Continue For
            End If

            Dim Field = New TableFields
            Field.Name = a.Item(0).ToString.Replace("""", "")
            Field.Type = a.Item(1).ToString.Replace("""", "")
            Field.CouldBeNull = a.Item(2).ToString
            Field.Length = a.Item(3).ToString
            Field.Precision = a.Item(4).ToString


            Dim FieldSize = Field.CouldBeNull

            If Field.Type = "B" Then
                FieldSize = FieldSize + Field.Length
            ElseIf Field.Type = "L" Then
                FieldSize = FieldSize + 1
            ElseIf Field.Type = "N" Then
                FieldSize = FieldSize + Math.Truncate((Field.Length + 2) / 2)
            ElseIf Field.Type = "NC" Then
                FieldSize = FieldSize + Field.Length * 2
            ElseIf Field.Type = "NVC" Then
                FieldSize = FieldSize + Field.Length * 2 + 2
            ElseIf Field.Type = "RV" Then
                FieldSize = FieldSize + 16
            ElseIf Field.Type = "I" Then
                FieldSize = FieldSize + 8
            ElseIf Field.Type = "T" Then
                FieldSize = FieldSize + 8
            ElseIf Field.Type = "DT" Then
                FieldSize = FieldSize + 7
            ElseIf Field.Type = "NT" Then
                FieldSize = FieldSize + 8
            End If

            Field.Size = FieldSize
            Field.Offset = RowSize

            RowSize = RowSize + FieldSize

            PageHeader.Fields.Add(Field)

        Next

        PageHeader.RowSize = RowSize

        '{"Files",118,119,96}
        'Данные, BLOB, индексы

        Dim BlockData = Convert.ToInt32(ParsedString.Item(0).Item(5).Item(1).ToString)
        Dim BlockBlob = Convert.ToInt32(ParsedString.Item(0).Item(5).Item(2).ToString)

        PageHeader.BlockData = BlockData
        PageHeader.BlockBlob = BlockBlob

    End Sub

End Module
